Added http client implementation based on Eiffel Net (i.e socket)

jocelyn opened the pull request.

colin-adams commented on library/network/http_client/src/default/libcurl_or_net/default_http_client.e:
```diff
@@ -0,0 +1,43 @@
+note
+	description: "[
+			Default HTTP_CLIENT based on LIBCURL_HTTP_CLIENT.
```
That sounds like a copy-and-past error.

jocelyn replied:
fixed

colin-adams commented on library/network/http_client/src/http_client_request.e:
```diff
@@ -14,25 +14,38 @@ inherit
 	context: detachable HTTP_CLIENT_REQUEST_CONTEXT
 
+	current_redirects: INTEGER_32
+
```
Where's the header comment for this attribute?

jocelyn replied:
fixed by removing it, since it was not needed.
I missed that when I did the initial review from student project.
Thanks.

colin-adams commented on library/network/http_client/src/spec/net/net_http_client_request.e:
```diff
@@ -0,0 +1,636 @@
+		rescue
+			retried := True
+			retry
+		end
```
That was a rather long routine.

jocelyn replied:
There are room for improvement, especially better design, provide component to build multipart/form-data content, and so on.
Given the time I had to work on this, I tried to make routine smaller, but there is still some work to do there.
I will try to find time, to improve that student work, and integrate it to EWF when ready.
For now, I shared this initial review and improvements on top of Florian Jacky code (from student project), to get feedback, and maybe contribution.
And also to show this new implementation is coming to reality

colin-adams commented on library/network/http_client/src/spec/net/net_http_client_request.e:
```diff
@@ -0,0 +1,636 @@
+					end
+				end
+			end
+		end
```
Another very long routine.

jocelyn replied:
I tried to reduce it, by extracting code. But still big.

colin-adams commented on library/network/http_client/src/spec/null/null_http_client.e:
```diff
@@ -0,0 +1,33 @@
+note
+	description : "[
+				Instantiate one of the descendant of HTTP_CLIENT
```
This sounds like it should be more specific. "one of the descendants"?

jocelyn replied:
fixed

colin-adams commented on library/network/http_client/src/spec/null/null_http_client_session.e:
```diff
@@ -0,0 +1,114 @@
+note
+	description : "[
+				HTTP_CLIENT_SESSION represents a session
```
Could be, but this is a description of the ancestor, surely.

jocelyn replied:
fixed

jocelyn commented:
I forgot to mention, but the code is the result of student project from Florian Jacky.
And I updated/fixed/improved part of the code. 
But there are still work to do.
For instance try to support persistent connection, support multiple upload files, and so on.
The libcurl implementation added constraints, but I guess we can remove them, and eventually report "not supported" feature .. when libcurl is used.
And/Or reverse.

I guess , for now, there is no support for proxy in net implementation.
It will come if we, or a user has the need.
Either by contribution by any user, or update from EWF team.

Regards, and thanks for any comment, feedback, code.

jocelyn commented:
Additional information:
this pull request is mostly to get feedback and code review.
this is not YET intended to be merged into main EWF.

There are still a few issues, maybe related to the use of  ready_for_reading , or various timeout.
investigation is in progress.
As a real test, we are trying to make it work with firebase rest api. (see EWF/redwood project)

Regards.

colin-adams commented:
How/where do I see EWF/redwood project?

jocelyn commented:
This is an experimental library to access firebase REST api, mostly contributed by Javier Velilla.
And this is good, since current "net" http_client has trouble with it, while the "libcurl" http_client seems to work fine.
So this is good opportunity to reproduce issue and hopefully fix them.

redwood, or the Eiffel firebase api : https://github.com/EiffelWebFramework/Redwood

jocelyn commented:
Thanks to Javier, we fixed the timeout issue.
And in the meantime, I also fixed a few protocol issues, and header related issue.

Now, next step if more tests, and obviously have smaller routine.
As TODO, we may have
- implement proxy support (would be helpful with fiddler2 for instance)
- implement better MIME encoding, for now the "boundary" is always the same .. which is not good.
- and a few others points that I may have forgetten for now.

BUT the good news is the Eiffel firebase library (named redwood), is working with the "net" http client.

thanks again Javier, that pointed me to what was under my nose, and helped fixing that code (I was blind for a time ;) )

jocelyn commented:
It was integrated long time ago, closing this pending PR.
